Patient oncompletes

thomasguerena opened the pull request:
Added "patient" oncompletes to backup, restore, and clean. What this means is that the oncomplete callback passed to backup/restore/clean doesn't get called until each of the sub-backups/restores/cleans have finished.

thomasguerena committed:
Added patient oncomplete handler to backup.

thomasguerena committed:
Added patient callbacks to restore functions.

thomasguerena committed:
Added patient callbacks to clean.

RyanLB commented on modules/media.js:
```diff
@@ -55,23 +55,29 @@ Media.prototype.initialize = function() {
-  }, oncomplete);
+  }, function(error) {
+    if (error || empty) {
+      oncomplete(type, error);
```
Does oncomplete ever get called if there's no error?

RyanLB replied:
Also, we call `oncomplete(type, error)` even in the case where `error` is undefined and `empty` is true.

RyanLB commented on modules/media.js:
```diff
@@ -119,11 +127,16 @@ Media.prototype.backup = function(type, oncomplete) {
+  ffosbr.media.get('sdcard1', paths[type], function(file, error) {
+
+    if (error || !file) {
+      return oncomplete(type, error);
```
Again, we'll call `oncomplete(type, error)` if `error` is undefined and `!file` evaluates to `true`

RyanLB commented:
`ffosbr.photos.restore()` invokes its callback repeatedly. My assumption is that the individual modules for media types haven't been updated for patient callbacks

thomasguerena committed:
Added callback chaining to media restore method and added a flag to the backup/restore/clean to ignore all further async activity after oncomplete has been called.

thomasguerena commented:
Latest commit fixed multiple oncompletes. Uses some recursive callback chaining and a "ignore flag" in the backup/restore/clean modules to filter out false signals fired after the oncomplete.
Ready for another review.

RyanLB commented:
I think media.375 should call `oncomplete`, not `forEach`

RyanLB commented:
The comment about the redundant check for `!file` isn't super important. If the comment about media.js 375 is resolved (or isn't actually a problem), this looks good to me

RyanLB commented on modules/media.js:
```diff
@@ -157,15 +183,41 @@ Media.prototype.restore = function(type, oncomplete) {
+  // Ignore any errors in "get". They will show up in the oncomplete.
+  ffosbr.media.get('sdcard1', paths[type], function(file, error) {
+    if (!file) {
+      empty = true;
```
Since `media.get()` checks to see whether `file` exists before entering `forEach`, this branch is never taken.

RyanLB commented on modules/media.js:
```diff
@@ -157,15 +183,41 @@ Media.prototype.restore = function(type, oncomplete) {
+  ffosbr.media.get('sdcard1', paths[type], function(file, error) {
+    if (!file) {
+      empty = true;
+    } else if (file) {
```
Since this `else if` is the opposite of the `if` condition, the `if` part is unnecessary.

RyanLB commented on modules/media.js:
```diff
@@ -157,15 +183,41 @@ Media.prototype.restore = function(type, oncomplete) {
+  }, function(error) {
+    if (error || empty) {
+      oncomplete(type, error);
+    } else {
```
`empty` is initialized to `true` and never set to `false`; this branch is never taken, and `restore()` does nothing

RyanLB committed:
Merge branch 'master' into patient-oncompletes

Conflicts:
	modules/settings.js

RyanLB committed:
Fix error that caused restore() to skip the actual restoration.
• Made media restoration store errors in an array instead of breaking the restoration chain
• Removed the "empty" variable in media restoration, instead handling this in writeFiles()
• Recomment music for now, that'll be a different branch

RyanLB committed:
Make Clean continue rolling and aggregate errors.
• Fix an error where get() calls forEach() instead of oncomplete() to report an error
• Make clean() ignore the empty variable
